FAPI Crashes in test_policy_action

I am on tpm2-tss version 3.1.0, Ubuntu 20.04, Python 3.8.5 and CFFI 1.14.5. I was initially thinking that this crash had something to do with Callback fn signature mismatches, but the testing TSS version is 3.0.3 which has those changes, so I am not sure.

```
================================================== test session starts ===================================================
platform linux -- Python 3.8.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/wcrobert/workspace/tpm2-pytss
plugins: cov-2.12.1, forked-1.3.0, xdist-2.2.1
gw0 [53] / gw1 [53] / gw2 [53] / gw3 [53]
................................................Fatal Python error: Aborted

Thread 0x00007f02a8ff3700 (most recent call first):
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 400 in read
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 432 in from_io
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 967 in _thread_receiver
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 220 in run
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 285 in _perform_spawn

Current thread 0x00007f02aa0b7740 (most recent call first):
  File "/home/wcrobert/workspace/tpm2-pytss/tpm2_pytss/FAPI.py", line 331 in sign
  File "/home/wcrobert/workspace/tpm2-pytss/test/test_fapi.py", line 900 in test_policy_action
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/xdist/remote.py", line 89 in run_one_test
  File "/home/wcrobert/.local/lib/python3.8/site-packages/xdist/remote.py", line 75 in pytest_runtestloop
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/xdist/remote.py", line 244 in <module>
  File "/home/wcrobert/.local/lib/python3.8/site-packa.ges/execnet/gateway_base.py", line 1084 in executetask
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 220 in run
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 285 in _perform_spawn
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 267 in integrate_as_primary_thread
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 1060 in serve
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 1554 in serve
  File "<string>", line 8 in <module>
  File "<string>", line 1 in <module>
...[gw2] node down: Not properly terminated
F
replacing crashed worker gw2
gw0 [53] / gw1 [53] / gw4 ok / gw3 [53]  
======================================================== FAILURES ========================================================
___________________________________________________ test/test_fapi.py ____________________________________________________
[gw2] linux -- Python 3.8.5 /usr/bin/python3
worker 'gw2' crashed while running 'test/test_fapi.py::TestFapi::test_policy_action'
================================================ short test summary info =================================================
FAILED test/test_fapi.py::TestFapi::test_policy_action
============================================== 1 failed, 52 passed in 3.72s ==============================================
```

--- a/docs/requirements.txt
+++ b/docs/requirements.txt
@@ -12,6 +12,7 @@ imagesize==1.2.0
 Jinja2==2.11.3
 MarkupSafe==1.1.1
 packaging==20.9
+pkgconfig==1.5.5
 pycparser==2.20
 Pygments==2.8.1
 pyparsing==2.4.7

--- a/test/test_fapi.py
+++ b/test/test_fapi.py
@@ -6,14 +6,13 @@ import binascii
 import random
 import string
 
-import pkgconfig
-
 import pytest
 from cryptography.hazmat.primitives import hashes, serialization
 from cryptography.hazmat.primitives.asymmetric import ec
 from cryptography.hazmat.backends import default_backend
 
 from tpm2_pytss import *
+from tpm2_pytss.utils import is_bug_fixed
 
 from .TSS2_BaseTest import TpmSimulator, TSS2_BaseTest
 
@@ -315,7 +314,7 @@ class TestFapi:
     # TODO test encrypt with RSA profile. Needs to be provisioned separately.
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2028"
+        not is_bug_fixed(fixed_in="3.2"), reason="tpm2-tss bug, see #2028"
     )
     def test_import_key_double_ok(self, cryptography_key):
         key, key_public_pem = cryptography_key
@@ -329,7 +328,7 @@ class TestFapi:
         assert imported is False
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2028"
+        not is_bug_fixed(fixed_in="3.2"), reason="tpm2-tss bug, see #2028"
     )
     def test_import_key_double_fail(self, cryptography_key):
         key, key_public_pem = cryptography_key
@@ -341,7 +340,7 @@ class TestFapi:
             self.fapi.import_object(path=key_path, import_data=key_public_pem)
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2028"
+        not is_bug_fixed(fixed_in="3.2"), reason="tpm2-tss bug, see #2028"
     )
     def test_import_policy_double_ok(self):
         policy = """
@@ -360,7 +359,7 @@ class TestFapi:
         assert imported is False
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2028"
+        not is_bug_fixed(fixed_in="3.2"), reason="tpm2-tss bug, see #2028"
     )
     def test_import_policy_double_fail(self):
         policy = """
@@ -593,7 +592,8 @@ class TestFapi:
         self.fapi.sign(key_path, b"\x22" * 32)
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2084"
+        not is_bug_fixed(fixed_in="3.2", backports=["2.4.7", "3.0.5", "3.1.1"]),
+        reason="tpm2-tss bug, see #2084",
     )
     def test_write_authorize_nv(self, esys):
         # write CommandCode policy for sign key into nv index
@@ -639,7 +639,8 @@ class TestFapi:
             self.fapi.quote(path=key_path, pcrs=[7, 9])
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2084"
+        not is_bug_fixed(fixed_in="3.2", backports=["2.4.7", "3.0.5", "3.1.1"]),
+        reason="tpm2-tss bug, see #2084",
     )
     def test_authorize_policy(self, sign_key):
         # create policy Authorize, which is satisfied via a signature by sign_key
@@ -703,7 +704,7 @@ class TestFapi:
             self.fapi.quote(path=key_path, pcrs=[7, 9])
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2080"
+        not is_bug_fixed(fixed_in="3.2"), reason="tpm2-tss bug, see #2080"
     )
     def test_policy_signed(self, cryptography_key):
         # create external signing key used by the signing authority external to the TPM
@@ -868,7 +869,8 @@ class TestFapi:
         self.fapi.delete(path=nv_path)
 
     @pytest.mark.skipif(
-        pkgconfig.installed("tss2-fapi", "<=3.1.0"), reason="tpm2-tss bug, see #2089"
+        not is_bug_fixed(fixed_in="3.2", backports=["2.4.7", "3.0.5", "3.1.1"]),
+        reason="tpm2-tss bug, see #2089",
     )
     def test_policy_action(self):
         # create policy Action, which is satisfied via the callback

--- a/tpm2_pytss/FAPI.py
+++ b/tpm2_pytss/FAPI.py
@@ -11,7 +11,7 @@ from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 from ._libtpm2_pytss import ffi, lib
 from .callbacks import Callback, CallbackType, get_callback, unlock_callback
 from .fapi_info import FapiInfo
-from .utils import _chkrc, to_bytes_or_null, get_dptr
+from .utils import _chkrc, check_bug_fixed, get_dptr, to_bytes_or_null
 from .TCTI import TCTI
 
 logger = logging.getLogger(__name__)
@@ -496,6 +496,10 @@ class FAPI:
         Returns:
             bool: True if the object was imported. False otherwise.
         """
+        check_bug_fixed(
+            fixed_in="3.2",
+            details="FAPI Import will overwrite existing objects with same path silently. See https://github.com/tpm2-software/tpm2-tss/issues/2028",
+        )
         path = to_bytes_or_null(path)
         import_data = to_bytes_or_null(import_data)
         ret = lib.Fapi_Import(self._ctx, path, import_data)
@@ -682,6 +686,12 @@ class FAPI:
         Returns:
             bytes: The platform certificates
         """
+        check_bug_fixed(
+            fixed_in="3.2",
+            backports=["2.4.7", "3.0.5", "3.1.1"],
+            details="FAPI Get Platform Certificate might lead wrong sequence errors. See https://github.com/tpm2-software/tpm2-tss/issues/2091",
+        )
+
         # TODO split certificates into list
         # TODO why bytes? is this DER?
         certificate = ffi.new("uint8_t **")
@@ -895,6 +905,11 @@ class FAPI:
         Returns:
             Tuple[str, bytes, str, str]: info, signature, pcr_log, certificate
         """
+        check_bug_fixed(
+            fixed_in="3.2",
+            backports=["2.4.7", "3.0.5", "3.1.1"],
+            details="Multiple calls of FAPI Quote might lead to TPM out of memory errors. See https://github.com/tpm2-software/tpm2-tss/issues/2084",
+        )
 
         path = to_bytes_or_null(path)
         quote_type = to_bytes_or_null(quote_type)
@@ -1262,6 +1277,11 @@ class FAPI:
         Raises:
             TSS2_Exception: If Fapi returned an error code.
         """
+        check_bug_fixed(
+            fixed_in="3.2",
+            details="FAPI PolicySigned default nameAlg might be SHA1 unexpectedly. See https://github.com/tpm2-software/tpm2-tss/issues/2080. Fixed in https://github.com/tpm2-software/tpm2-tss/commit/b843960b6e601a786b469832392dc0a12e13cf34",
+        )
+
         if callback is None and user_data is not None:
             raise RuntimeError("If callback is None, user_data must be None, too.")
 
@@ -1334,6 +1354,12 @@ class FAPI:
         """
         if callback is None and user_data is not None:
             raise ValueError("If callback is None, user_data must be None, too.")
+
+        check_bug_fixed(
+            fixed_in="3.2",
+            backports=["2.4.7", "3.0.5", "3.1.1"],
+            details="FAPI Policy Action might lead to crashes. See https://github.com/tpm2-software/tpm2-tss/issues/2089",
+        )
 
         if user_data is None:
             user_data_len = 0

--- a/tpm2_pytss/utils.py
+++ b/tpm2_pytss/utils.py
@@ -1,11 +1,16 @@
 """
 SPDX-License-Identifier: BSD-2
 """
-
+import logging
 import sys
+from typing import List
+
+import pkgconfig
 
 from ._libtpm2_pytss import ffi, lib
 from .TSS2_Exception import TSS2_Exception
+
+logger = logging.getLogger(__name__)
 
 # Peek into the loaded modules, if mock is loaded, set __MOCK__ to True, else False
 __MOCK__ = "unittest.mock" in sys.modules
@@ -178,3 +183,43 @@ def check_friendly_int(friendly, varname, clazz):
         raise ValueError(
             f"expected {varname} value of {friendly} in class {str(clazz)}, however it's not found."
         )
+
+
+def is_bug_fixed(
+    fixed_in=None, backports: List[str] = None, lib: str = "tss2-fapi"
+) -> bool:
+    """Use pkg-config to determine if a bug was fixed in the currently installed tpm2-tss version."""
+    if fixed_in and pkgconfig.installed(lib, f">= {fixed_in}"):
+        return True
+
+    version = pkgconfig.modversion(lib)
+    version = version.split("-")[0]
+    vers_major, vers_minor, vers_patch = (int(s) for s in version.split("."))
+
+    if backports is None:
+        backports = []
+    for backport in backports:
+        backp_major, backp_minor, backp_patch = (int(s) for s in backport.split("."))
+
+        if vers_major == backp_major and vers_minor == backp_minor:
+            return vers_patch >= backp_patch
+
+    return False
+
+
+def check_bug_fixed(
+    details,
+    fixed_in=None,
+    backports: List[str] = None,
+    lib: str = "tss2-fapi",
+    error: bool = False,
+) -> None:
+    """Emit a warning or exception if there is an unfixed bug in the currently installed tpm2-tss version."""
+    if not is_bug_fixed(fixed_in=fixed_in, backports=backports, lib=lib):
+        version = pkgconfig.modversion(lib)
+        message = f"This API call {'is' if error else 'may be'} affected by a bug in {lib} version {version}: {details}\nPlease use >= {fixed_in}. Backports exist for {backports}."
+
+        if error:
+            raise RuntimeError(message)
+
+        logger.warning(message)

